### PR TITLE
[dynamo] Track params/buffers and mark them as static

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132334

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -1916,7 +1917,7 @@ def TENSOR_MATCH(self, guard: Guard, value=None):
             #
             assert guard.source is not None
             static, reason = tensor_always_has_static_shape(
-                value, is_tensor=True, guard_source=guard.source
+                value, is_tensor=True, tensor_source=guard.originating_source
```
Comment (ezyang) [inline torch/_dynamo/guards.py:1920]: What's going on here?

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -2156,7 +2156,7 @@ def tensor_static_reason_to_message(reason: TensorStaticReason):
 def tensor_always_has_static_shape(
     tensor: Union[torch.Tensor, Any],
     is_tensor: bool,
-    guard_source: "torch._guards.GuardSource",
+    tensor_source: Any,
```
Comment (ezyang) [inline torch/_dynamo/utils.py:2159]: Why did this annotation get worse?

**Review by ezyang — APPROVED** [review]
Seems alright I guess. I feel like you can probably infer it from the Source without needing a separate dataclass

File: torch/_dynamo/guards.py (hunk 1 above)
Comment (anijain2305) [reply torch/_dynamo/guards.py:1920]: Ah, forgot to fix it.

File: torch/_dynamo/utils.py (hunk 2 above)
Comment (anijain2305) [reply torch/_dynamo/utils.py:2159]: Just change in the argument. It was a guard_source earlier (which is a specific thing and of type GuardSource). Now it's just source.

Comment (ezyang) [reply torch/_dynamo/utils.py:2159]: well we've got a perfectly good Source type we can put here :)

File: test/inductor/test_cpu_select_algorithm.py [hunk 3]
```diff
@@ -560,6 +560,7 @@ def forward(self, x, other, other2):
 
     @inductor_config.patch({"freezing": True})
     @patches
+    @torch._dynamo.config.patch(inline_inbuilt_nn_modules=True)
```
Comment (leslie-fang-intel) [inline test/inductor/test_cpu_select_algorithm.py:563]: Perhaps change the `inline_inbuilt_nn_modules` in the patch to make it effects for all the UT in this test file.

**Review by mlazos — APPROVED** [review]
Looks good, this might be useful for marking buffers/params as static for cudagraphs as well, but I haven't seen any issue surface with marking them at nn module variable construction time.

**Comment by anijain2305** [comment]
Relanding at https://github.com/pytorch/pytorch/pull/132538

**Pull request closed without merging**